File: extension/test/contract.test.ts
import assert from 'node:assert/strict';
import { readFileSync } from 'node:fs';
import { fileURLToPath } from 'node:url';
import { test } from 'node:test';

import { mask, unmask, type AgentConfig } from '../src/agent.ts';

// The extension's half of the contract with the agent.
//
// The same file internal/proxy/contract_test.go replays against a real agent. That
// one proves the agent answers these responses; this one proves the client sends
// these requests and reads those answers. Neither is enough alone, and together they
// are what the same-repository decision was made for: a field renamed on either side
// fails on the other, in the commit that renamed it.
//
// Nothing here is generated. The requests and the reasons in the file are written by
// hand; only the responses are recorded, and only by `make contract-update`.

type Exchange = {
  name: string;
  why: string;
  route: string;
  request: Record<string, unknown>;
  response: Record<string, unknown>;
};

type Contract = {
  detector: { locales: string[]; substitution: string; secret_level: string };
  session: string;
  exchanges: Exchange[];
};

const contract = JSON.parse(
  readFileSync(fileURLToPath(new URL('../testdata/contract.json', import.meta.url)), 'utf8'),
) as Contract;

const cfg: AgentConfig = {
  baseUrl: 'http://127.0.0.1:8787',
  key: '0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef',
};

/** recording is a fetch that answers one recorded exchange and keeps what it was
 * asked, so the request the client actually built can be compared to the file. */
function recording(exchange: Exchange) {
  const seen: { url?: string; init?: RequestInit } = {};
  const fetchImpl = async (url: string, init?: RequestInit): Promise<Response> => {
    seen.url = url;
    seen.init = init;
    return new Response(JSON.stringify(exchange.response), {
      status: 200,
      headers: { 'Content-Type': 'application/json' },
    });
  };
  return { seen, fetchImpl };
}

for (const exchange of contract.exchanges) {
  test(`contract: ${exchange.name}`, async () => {
    const { seen, fetchImpl } = recording(exchange);

    let result: unknown;
    if (exchange.route === '/mask') {
      result = await mask(cfg, contract.session, exchange.request.texts as string[], fetchImpl);
    } else if (exchange.route === '/unmask') {
      result = await unmask(
        cfg,
        contract.session,
        {
          text: exchange.request.text as string,
          tail: exchange.request.tail as string,
          final: exchange.request.final as boolean,
        },
        fetchImpl,
      );
    } else {
      throw new Error(`the contract names a route this client has no call for: ${exchange.route}`);
    }

    assert.equal(seen.url, cfg.baseUrl + exchange.route, exchange.why);
    assert.equal(seen.init?.method, 'POST');

    // The request the client built, against the one in the file. This is the half
    // that catches a client which quietly stopped sending the tail: it would still
    // parse every recorded answer and restore nothing in a real stream.
    assert.deepEqual(JSON.parse(seen.init?.body as string), exchange.request, exchange.why);

    // And the answer, parsed into what the caller gets.
    assert.deepEqual(result, exchange.response, exchange.why);
  });
}

test('contract: every request carries the key and the session', async () => {
  const exchange = contract.exchanges[0]!;
  const { seen, fetchImpl } = recording(exchange);
  await mask(cfg, contract.session, exchange.request.texts as string[], fetchImpl);

  const headers = seen.init?.headers as Record<string, string>;
  assert.equal(headers['X-Cloakfleet-Control'], cfg.key,
    'without the control header the agent refuses, and the send is blocked');
  assert.equal(headers['X-Session-Id'], contract.session,
    '/mask and /unmask for one conversation must name the same session, or the expansion finds nothing');
});

test('contract: the session is one name for both routes', () => {
  // Held here rather than left to a reader noticing: the file has one session for
  // every exchange because that is the property, not a convenience of the fixture.
  assert.ok(contract.session.length > 0);
  assert.ok(
    contract.exchanges.length >= 2,
    'a contract with one exchange cannot show a mapping being carried across calls',
  );
});

File: internal/proxy/contract_test.go
package proxy

import (
	"encoding/json"
	"flag"
	"net/http/httptest"
	"os"
	"path/filepath"
	"reflect"
	"testing"

	"github.com/cloakfleet/cloakfleet/internal/detector"
	"github.com/cloakfleet/cloakfleet/internal/vault"
)

// The contract between this agent and the browser extension, as one file both sides
// read.
//
// This is the test the same-repository decision exists to make possible. The
// extension and the agent speak a protocol; in two repositories that protocol drifts,
// which is the class of failure this project documents everywhere else. Here the
// exchanges are recorded once, in a file the extension's own suite reads to drive its
// client and this one reads to drive the real agent — so a field renamed on either
// side fails on the other, in the same commit.
//
// The two halves check different things and neither is enough alone:
//
//   - this one: given the recorded request, the real detector and the real vault
//     answer the recorded response, exactly;
//   - the extension's: given the recorded response, its client parses what it
//     expects, and given the same inputs it produces the recorded request, exactly.
//
// The file is regenerated with `make contract-update`, deliberately as awkward as
// `make score-update`: a fixture that rewrites itself on every run agrees with
// whatever the code does, including a field it silently stopped sending.

const contractPath = "../../extension/testdata/contract.json"

var updateContract = flag.Bool("update-contract", false,
	"rewrite the recorded exchanges from this run — explain the delta in the PR")

type contractFile struct {
	Comment   string             `json:"comment"`
	Detector  contractDetector   `json:"detector"`
	Session   string             `json:"session"`
	Exchanges []contractExchange `json:"exchanges"`
}

// contractDetector pins what the agent must be configured as for the recorded
// answers to be the right ones.
//
// Recorded rather than assumed, because every one of them changes what comes back: a
// different locale set recognises different values, and fake mode replaces them with
// prose rather than tokens.
type contractDetector struct {
	Locales      []string `json:"locales"`
	Substitution string   `json:"substitution"`
	SecretLevel  string   `json:"secret_level"`
}

type contractExchange struct {
	Name string `json:"name"`

	// Why is what this exchange is here to hold. Written by hand and never
	// regenerated: an exchange nobody can say the purpose of is one somebody will
	// delete the next time it fails.
	Why string `json:"why"`

	Route    string          `json:"route"`
	Request  json.RawMessage `json:"request"`
	Response json.RawMessage `json:"response"`
}

// TestExtensionContract replays every recorded exchange against a real agent.
//
// One agent for the whole file, in order: the exchanges build on each other, because
// that is what a conversation does. The token an /unmask exchange expands is the one
// an earlier /mask exchange minted, and replaying them independently would test a
// protocol nothing speaks.
func TestExtensionContract(t *testing.T) {
	file := loadContract(t)

	det := detector.New(detector.Config{Locales: file.Detector.Locales})
	mode, err := detector.ParseSubstitution(file.Detector.Substitution)
	if err != nil {
		t.Fatal(err)
	}
	det.SetSubstitution(mode)
	level, err := detector.ParseSecretLevel(file.Detector.SecretLevel)
	if err != nil {
		t.Fatal(err)
	}
	det.SetSecretLevel(level)

	v, err := vault.New(vault.NewMemory(), nil, vault.DefaultTTL)
	if err != nil {
		t.Fatal(err)
	}
	srv, err := New(Config{
		Providers:  []Provider{{Code: "anthropic", BaseURL: "https://example.invalid"}},
		ControlKey: testControlKey,
	}, det, v)
	if err != nil {
		t.Fatal(err)
	}
	agent := httptest.NewServer(srv.Handler())
	t.Cleanup(agent.Close)

	live := make([]json.RawMessage, len(file.Exchanges))
	for i, exchange := range file.Exchanges {
		got := postAs(t, agent, exchange.Route, testControlKey, file.Session, string(exchange.Request))
		if got.status != 200 {
			t.Fatalf("%s: status %d: %s", exchange.Name, got.status, got.body)
		}
		live[i] = json.RawMessage(got.body)

		if *updateContract {
			continue
		}
		if !sameJSON(t, exchange.Response, live[i]) {
			t.Errorf("%s\n  why:  %s\n  sent: %s\n  want: %s\n  got:  %s",
				exchange.Name, exchange.Why, exchange.Request, exchange.Response, got.body)
		}
	}

	if *updateContract {
		writeContract(t, file, live)
	}
}

// TestContractCoversWhatMatters holds the file itself to a standard.
//
// A contract file is only worth what it exercises, and it is the kind of fixture that
// quietly loses a case: somebody deletes a failing exchange, regenerates, and the
// suite goes green over a protocol nobody checks any more. So the shapes that must be
// in it are named here rather than left to whoever last edited the JSON.
func TestContractCoversWhatMatters(t *testing.T) {
	file := loadContract(t)

	seen := map[string]bool{}
	for _, e := range file.Exchanges {
		if e.Why == "" {
			t.Errorf("exchange %q says nothing about what it holds", e.Name)
		}
		seen[e.Route] = true
	}
	for _, route := range []string{"/mask", "/unmask"} {
		if !seen[route] {
			t.Errorf("nothing in the contract exercises %s", route)
		}
	}

	// A split replacement and a final flush: the two cases the tail exists for, and
	// the two a client would pass without if it simply never carried one.
	var carriesTail, isFinal bool
	for _, e := range file.Exchanges {
		if e.Route != "/unmask" {
			continue
		}
		var req unmaskRequest
		if err := json.Unmarshal(e.Request, &req); err != nil {
			t.Fatalf("%s: %v", e.Name, err)
		}
		if req.Tail != "" {
			carriesTail = true
		}
		if req.Final {
			isFinal = true
		}
	}
	if !carriesTail {
		t.Error("no recorded exchange carries a tail, so the split-replacement case is untested")
	}
	if !isFinal {
		t.Error("no recorded exchange ends the stream, so the flush is untested")
	}
}

func loadContract(t *testing.T) contractFile {
	t.Helper()

	raw, err := os.ReadFile(filepath.Clean(contractPath))
	if err != nil {
		t.Fatal(err)
	}
	var file contractFile
	if err := json.Unmarshal(raw, &file); err != nil {
		t.Fatalf("%s: %v", contractPath, err)
	}
	return file
}

func writeContract(t *testing.T, file contractFile, live []json.RawMessage) {
	t.Helper()

	for i := range file.Exchanges {
		compact := &json.RawMessage{}
		if err := json.Unmarshal(live[i], compact); err != nil {
			t.Fatal(err)
		}
		file.Exchanges[i].Response = *compact
	}

	encoded, err := json.MarshalIndent(file, "", "  ")
	if err != nil {
		t.Fatal(err)
	}
	if err := os.WriteFile(contractPath, append(encoded, '\n'), 0o600); err != nil {
		t.Fatal(err)
	}
	t.Logf("rewrote %s from this run", contractPath)
}

// sameJSON compares two documents by value rather than by bytes, so whitespace and
// key order in the file are not part of the contract. What is in it — a field
// present, a field absent, a value — is.
func sameJSON(t *testing.T, want, got json.RawMessage) bool {
	t.Helper()

	var a, b any
	if err := json.Unmarshal(want, &a); err != nil {
		t.Fatalf("the recorded response is not JSON: %v", err)
	}
	if err := json.Unmarshal(got, &b); err != nil {
		t.Fatalf("the agent answered something that is not JSON: %v", err)
	}
	return reflect.DeepEqual(a, b)
}
